File: kukuri-tauri/src/lib/api/p2p.ts
import { invokeCommand, invokeCommandVoid } from '@/lib/api/tauriClient';

export interface P2PStatus {
  connected: boolean;
  endpoint_id: string;
  active_topics: TopicStatus[];
  peer_count: number;
  metrics_summary: GossipMetricsSummary;
}

export interface TopicStatus {
  topic_id: string;
  peer_count: number;
  message_count: number;
  last_activity: number;
}

export interface GossipMetricDetails {
  total: number;
  failures: number;
  last_success_ms: number | null;
  last_failure_ms: number | null;
}

export interface GossipMetricsSummary {
  joins: number;
  leaves: number;
  broadcasts_sent: number;
  messages_received: number;
}

export interface GossipMetricsSection {
  joins: number;
  leaves: number;
  broadcasts_sent: number;
  messages_received: number;
  join_details: GossipMetricDetails;
  leave_details: GossipMetricDetails;
  broadcast_details: GossipMetricDetails;
  receive_details: GossipMetricDetails;
}

export interface MainlineMetrics {
  connected_peers: number;
  connection_attempts: number;
  connection_successes: number;
  connection_failures: number;
  connection_last_success_ms: number | null;
  connection_last_failure_ms: number | null;
  routing_attempts: number;
  routing_successes: number;
  routing_failures: number;
  routing_success_rate: number;
  routing_last_success_ms: number | null;
  routing_last_failure_ms: number | null;
  reconnect_attempts: number;
  reconnect_successes: number;
  reconnect_failures: number;
  last_reconnect_success_ms: number | null;
  last_reconnect_failure_ms: number | null;
}

export interface P2PMetrics {
  gossip: GossipMetricsSection;
  mainline: MainlineMetrics;
}

export interface BootstrapConfig {
  mode: 'default' | 'custom';
  nodes: string[];
}

export const p2pApi = {
  /**
   * P2P讖溯・繧貞・譛溷喧
   */
  initialize: () => invokeCommandVoid('initialize_p2p'),

  /**
   * 繝医ヴ繝・け縺ｫ蜿ょ刈
   */
  joinTopic: (topicId: string, initialPeers: string[] = []) =>
    invokeCommandVoid('join_p2p_topic', { topicId, initialPeers }),

  /**
   * 繝医ヴ繝・け蜷阪〒蜿ょ刈
   */
  joinTopicByName: (topicName: string, initialPeers: string[] = []) =>
    invokeCommandVoid('join_topic_by_name', { topicName, initialPeers }),

  /**
   * 繝医ヴ繝・け縺九ｉ髮｢閼ｱ
   */
  leaveTopic: (topicId: string) => invokeCommandVoid('leave_p2p_topic', { topicId }),

  /**
   * 繝医ヴ繝・け縺ｫ繝｡繝・そ繝ｼ繧ｸ繧偵ヶ繝ｭ繝ｼ繝峨く繝｣繧ｹ繝・   */
  broadcast: (topicId: string, content: string) =>
    invokeCommandVoid('broadcast_to_topic', { topicId, content }),

  /**
   * P2P謗･邯夂憾諷九ｒ蜿門ｾ・   */
  getStatus: () => invokeCommand<P2PStatus>('get_p2p_status'),

  /**
   * 繝弱・繝峨い繝峨Ξ繧ｹ繧貞叙蠕・   */
  getNodeAddress: async () => {
    const response = await invokeCommand<{ addresses: string[] }>('get_node_address');
    return response.addresses;
  },

  /**
   * 謖・ｮ壹＆繧後◆繝斐い繧｢繝峨Ξ繧ｹ縺ｫ謇句虚縺ｧ謗･邯・   */
  connectToPeer: (peerAddress: string) =>
    invokeCommandVoid('connect_to_peer', { peerAddress }),

  // Bootstrap UI
  getBootstrapConfig: () => invokeCommand<BootstrapConfig>('get_bootstrap_config'),
  setBootstrapNodes: (nodes: string[]) => invokeCommandVoid('set_bootstrap_nodes', { nodes }),
  clearBootstrapNodes: () => invokeCommandVoid('clear_bootstrap_nodes'),

  /**
   * Gossip繝｡繝医Μ繧ｯ繧ｹ繧貞叙蠕・   */
  getMetrics: () => invokeCommand<P2PMetrics>('get_p2p_metrics'),
};

File: kukuri-tauri/src-tauri/src/presentation/dto/p2p.rs
use crate::application::services::p2p_service::{
    ConnectionStatus as ServiceConnectionStatus, PeerStatus as ServicePeerStatus,
};
use crate::presentation::dto::Validate;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionStatusResponse {
    Connected,
    Connecting,
    Disconnected,
    Error,
}

impl From<ServiceConnectionStatus> for ConnectionStatusResponse {
    fn from(value: ServiceConnectionStatus) -> Self {
        match value {
            ServiceConnectionStatus::Connected => ConnectionStatusResponse::Connected,
            ServiceConnectionStatus::Connecting => ConnectionStatusResponse::Connecting,
            ServiceConnectionStatus::Disconnected => ConnectionStatusResponse::Disconnected,
            ServiceConnectionStatus::Error => ConnectionStatusResponse::Error,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeerStatusResponse {
    pub node_id: String,
    pub address: String,
    pub connected_at: i64,
    pub last_seen: i64,
}

impl From<ServicePeerStatus> for PeerStatusResponse {
    fn from(value: ServicePeerStatus) -> Self {
        Self {
            node_id: value.node_id,
            address: value.address,
            connected_at: value.connected_at,
            last_seen: value.last_seen,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct P2PStatusResponse {
    pub connected: bool,
    pub connection_status: ConnectionStatusResponse,
    pub endpoint_id: String,
    pub active_topics: Vec<TopicStatus>,
    pub peer_count: usize,
    pub peers: Vec<PeerStatusResponse>,
    pub metrics_summary: GossipMetricsSummaryResponse,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopicStatus {
    pub topic_id: String,
    pub peer_count: usize,
    pub message_count: usize,
    pub last_activity: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JoinTopicRequest {
    pub topic_id: String,
    pub initial_peers: Vec<String>,
}

impl Validate for JoinTopicRequest {
    fn validate(&self) -> Result<(), String> {
        if self.topic_id.is_empty() {
            return Err("Topic ID is required".to_string());
        }
        // 初期ピアのフォーマット検証は省略（実際には必要に応じて追加）
        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LeaveTopicRequest {
    pub topic_id: String,
}

impl Validate for LeaveTopicRequest {
    fn validate(&self) -> Result<(), String> {
        if self.topic_id.is_empty() {
            return Err("Topic ID is required".to_string());
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BroadcastRequest {
    pub topic_id: String,
    pub content: String,
}

impl Validate for BroadcastRequest {
    fn validate(&self) -> Result<(), String> {
        if self.topic_id.is_empty() {
            return Err("Topic ID is required".to_string());
        }
        if self.content.is_empty() {
            return Err("Content cannot be empty".to_string());
        }
        if self.content.len() > 50000 {
            return Err("Content is too large (max 50000 bytes)".to_string());
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NodeAddressResponse {
    pub addresses: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BootstrapConfigResponse {
    pub mode: String,
    pub nodes: Vec<String>,
    pub effective_nodes: Vec<String>,
    pub source: String,
    pub env_locked: bool,
    #[serde(default)]
    pub cli_nodes: Vec<String>,
    #[serde(default)]
    pub cli_updated_at_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayStatusResponse {
    pub url: String,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GossipMetricsResponse {
    pub joins: u64,
    pub leaves: u64,
    pub broadcasts_sent: u64,
    pub messages_received: u64,
    pub join_details: GossipMetricDetailsResponse,
    pub leave_details: GossipMetricDetailsResponse,
    pub broadcast_details: GossipMetricDetailsResponse,
    pub receive_details: GossipMetricDetailsResponse,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GossipMetricDetailsResponse {
    pub total: u64,
    pub failures: u64,
    pub last_success_ms: Option<u64>,
    pub last_failure_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MainlineMetricsResponse {
    pub connected_peers: u64,
    pub connection_attempts: u64,
    pub connection_successes: u64,
    pub connection_failures: u64,
    pub connection_last_success_ms: Option<u64>,
    pub connection_last_failure_ms: Option<u64>,
    pub routing_attempts: u64,
    pub routing_successes: u64,
    pub routing_failures: u64,
    pub routing_success_rate: f64,
    pub routing_last_success_ms: Option<u64>,
    pub routing_last_failure_ms: Option<u64>,
    pub reconnect_attempts: u64,
    pub reconnect_successes: u64,
    pub reconnect_failures: u64,
    pub last_reconnect_success_ms: Option<u64>,
    pub last_reconnect_failure_ms: Option<u64>,
    pub bootstrap: BootstrapMetricsResponse,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct P2PMetricsResponse {
    pub gossip: GossipMetricsResponse,
    pub mainline: MainlineMetricsResponse,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BootstrapMetricsResponse {
    pub env_uses: u64,
    pub user_uses: u64,
    pub bundle_uses: u64,
    pub fallback_uses: u64,
    pub last_source: Option<String>,
    pub last_applied_ms: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GossipMetricsSummaryResponse {
    pub joins: u64,
    pub leaves: u64,
    pub broadcasts_sent: u64,
    pub messages_received: u64,
}
